### PR TITLE
Support version annotation on any field. Deletes respect version.

### PR DESCRIPTION
**Pull request by Blackbaud-EricSlater**
- support @version annotation on any string field. 
- delete respects version

### CODE REVIEW

**Comment by Blackbaud-EricSlater** [comment]
I still have a bit of work to do here in adding similar changes to ReactiveCosmosTemplate. I will get that done ASAP.

I wanted to go ahead and open this PR since there is an implementation detail that needs to be discussed.

As it is, this will allow you to set @version on any string field. However, the entity in the database will save this field alongside the _etag field. This means that the actual document in the database will have two fields with the version value on it. The one that you annotated and the _etag field that cosmos manages

One solution to this is to remove the field marked with @version from the serialized json that we send to cosmos and then map that _etag field we get back from cosmos to to annotated field. The problem with this solution is that if users look in the database they will not see the field that they annotated with @version but instead they will see the field _etag.

Ideally cosmos would allow us to tell it the field to use for versioning but as far as i could tell from the docs this is not something that is supported. Are you guys aware of anything that would allow us to do this? If not I think we will have to go with the solution above, or leave it like it is and have both the @version annotated field and the system managed _etag field.

Personally, i dont like having both fields exist and and think that removing the annotated field from the serialized object is preferable as long as we document that the field in the database will be named _etag regardless and not what you mark with @version.

@kushagraThapar

**Comment by Blackbaud-EricSlater** [comment]
also, there are no breaking changes in this as we thought originally.

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - `_etag` is the field that is automatically generated by cosmosdb to honor the optimistic locking control mechanism. So I don't think you can tell Cosmos to choose any other field as this is integrated at the database level. 

Here is more information on it : https://docs.microsoft.com/en-us/azure/cosmos-db/database-transactions-optimistic-concurrency#optimistic-concurrency-control

And this answer : https://stackoverflow.com/a/49376962/2674287

Regarding your above solution, I agree and personally think we should NOT have two separate fields (one annotated with `@Version` tag and `_etag` field) and only keep `_etag` field in the database. However, this might create confusion for the general end user, who when looks up the documents in portal, will not see their field. And this might create unnecessary customer reported issues on our end (SDK team, or me :P ) 
So I think best way to go forward with this would be to keep both the fields in the database, and map `_etag` value to the `@Version` tag specified field. 
When deserializing back, we can just put the `_etag` field value to the `@Version` tag field. 
But let's keep both the values in the database for more clarity. 

Let me know if this makes sense.

**Comment by Blackbaud-EricSlater** [comment]
rebased on latest master. Reactive changes to follow shortly

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - One additional thought on the PR - I see on `EtagIT` you are re-using `Person` entity. 

I think since maven and SpringJUnit4ClassRunner runs some of these tests in parallel, there is a high possibility of facing issues when one integration test class is running and the other one removes the container (Person) in this case. 

Instead of using `PersonRepository` in `EtagIT`, can you create a new repository and new entity for this ?
Something like - `EtagTestEntity` and `EtagRepository` and use them in `EtagIT` 

Below are the errors I am seeing in the integration test run logs : 

```
2020-08-03T20:01:36.8295211Z [ERROR] Errors: 
2020-08-03T20:01:36.8300365Z [ERROR]   ReactiveCosmosTemplateIT.testOptimisticLockWhenUpdatingWithWrongEtag:251 ï¿½ NullPointer
2020-08-03T20:01:36.8301804Z [ERROR]   ReactiveCosmosTemplateIT.testUpsert:235 ï¿½ NullPointer
2020-08-03T20:01:36.8303048Z [ERROR]   ReactiveCosmosTemplateIT.testUpsertBySecondaryKey:275 ï¿½ NullPointer
2020-08-03T20:01:36.8304700Z [ERROR]   ReactiveCosmosTemplateIT.testUpsertWithContainerName:289 ï¿½ NullPointer
```

See how `ReactiveCosmosTemplateIT` is facing issues because `EtagIT` ran before it and cleaned up common resources which caused the `ReactiveCosmosTemplateIT` to break.

**Comment by Blackbaud-EricSlater** [comment]
reactive changes up. Ill refactor the tests as described. I will need to do the same for the reactive tests i added

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - thank you so much for bringing these changes in, they look good to me. 
Can you please fix the tests and push the changes ?

We have some big architectural changes coming up soon, which I am currently blocking so we can get your PR in first. 
Let me know your thoughts.

**Comment by Blackbaud-EricSlater** [comment]
@kushagraThapar test fixes are in. Sorry for the delay

File: sdk/cosmos/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/core/CosmosTemplateIT.java [hunk 1]
```diff
@@ -199,22 +199,19 @@ public void testUpsertNewDocument() {
     public void testUpdateWithReturnEntity() {
         final Person updated = new Person(TEST_PERSON.getId(), UPDATED_FIRST_NAME,
             TEST_PERSON.getLastName(), TEST_PERSON.getHobbies(), TEST_PERSON.getShippingAddresses());
-        updated.set_etag(insertedPerson.get_etag());
```
Comment (Blackbaud-EricSlater) [inline sdk/cosmos/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/core/CosmosTemplateIT.java:202]: oops this should not have been removed. Forgot i had renamed this field. Ill add this back

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater  perfect. 
No need to apologize :) 
You have made a great contribution with this feature, thanks again!
Once the CI passes, I will merge the PR in.

File: sdk/cosmos/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/core/CosmosTemplateIT.java (hunk 1 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/core/CosmosTemplateIT.java:202]: Yes, please add these asserts back.

Comment (Blackbaud-EricSlater) [reply sdk/cosmos/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/core/CosmosTemplateIT.java]: fixed

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - thanks again for the contribution.

**Comment by Blackbaud-MikeLueders** [comment]
@kushagraThapar just so we're all clear, the additional field in the database will never be correct.  That is, let's say the field was named `version`, a saved document would have `_etag` and `version` but the values in cosmos will never be the same b/c cosmos will automatically update the `_etag` value on update but it will not update the `version` field.  This means, a user looking at the `version` field in the application will never see that value in the database b/c the value of `version` in the application object will be the value of the cosmos `_etag` while the value of `version` in cosmos will be the previous value of `_etag`.

This has the potential to lead to more confusion.  Since `_etag` is cosmos's mechanism for optimistic locking while `@Version` is spring data's, it seems natural to me that the `@Version` field represents the `_etag` field in cosmos.

Either way you choose to go, I think it's worth adding a section note in the docs.

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - make sense, thanks for the explanation. It is definitely worth mentioning in the docs. I will update the readme for this. 
I will also have another discussion on this with my internal team and if it will make sense to remove the extra field that you mentioned at first place. That change should be easy to do if everyone agrees on that.

**Comment by Blackbaud-EricSlater** [comment]
@kushagraThapar just to elaborate on @Blackbaud-MikeLueders point, i think this makes a strong case for simply mapping between the @version field on the java object and the _etag field on the document and not having the extra field. You can imagine that hassle and confusion this could cause to other users of the document that are not using java. There would be this extra version field on the object (that is always wrong) that was stuck there by the java framework and that all other  non java consumers of the document need to ignore. 

it also adds to the size of the document which in large scale could effect cost

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater @Blackbaud-MikeLueders - sure, I will consider it and will try to get rid of the extra field from the Cosmos Document on the db side. 
I didn't realize it will be always wrong. That would definitely create more confusion.. Let me work on that change.. thanks again!

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - if it can be quick, please do so.. Since as I mentioned that I was blocking a big structural change because of your PR, I kind of promised to not do any more code changes until that big change happens :) 

But if you can create a PR now and it is in before tonight, I think that will be perfect. 
Thank you so much!

**Comment by Blackbaud-EricSlater** [comment]
@kushagraThapar will have it ready shortly.

**Pull request merged**